crates/ratchet-core/src/gpu/wgsl/access_granularity.rs

```diff
@@ -0,0 +1,34 @@
+use super::dtype::WgslDType;
+
+pub struct Vec4;
+pub struct Vec2;
+pub struct Scalar;
+
+pub trait AccessGranularity {
+    fn as_size(&self) -> usize;
+    fn render(&self, dtype: WgslDType) -> String {
+        let size = self.as_size();
+        match size {
+            4 => format!("vec4<{}>", dtype),
+            2 => format!("vec2<{}>", dtype),
+            1 => format!("{}", dtype),
+            _ => panic!("Invalid access granularity"),
+        }
+    }
+}
```

Comment:
Size can be a const generic on the trait.

If you then connect `AccessGranularity` and `WgslDType`at type level then the render result can be known at const time as well.

Reply:
I thought about this, reason I didn't:
https://willcrichton.net/rust-api-type-patterns/main_idea.html#principles-of-representation-cardinality-mismatch

Someone could pass in 5 and it would be valid if we had a const generic.

Yeah something like connecting the 2 at type level would be nice!

Reply:
Right but for ratchet that user would be you or the other contributors.